AutoCollection/HeartBeat.ts

```diff
@@ -0,0 +1,138 @@
+        return HeartBeat.INSTANCE && HeartBeat.INSTANCE._isEnabled;
+    }
+
+    public trackHeartBeat(config: Config, callback: () => void) {
+        let waiting: boolean = false;
+        let properties: {[key: string]: string} = {};
+        const sdkVersion = Context.sdkVersion; // "node" or "node-nativeperf"
+        properties["sdk"] = sdkVersion;
+        properties["osType"] = os.type();
+        if (process.env.WEBSITE_SITE_NAME) { // Web apps
+            properties["appSrv_SiteName"] = process.env.WEBSITE_SITE_NAME || "";
+            properties["appSrv_wsStamp"] = process.env.WEBSITE_HOME_STAMPNAME || "";
+            properties["appSrv_wsHost"] = process.env.WEBSITE_HOSTNAME || "";
+        } else if (process.env.FUNCTIONS_WORKER_RUNTIME) { // Function apps
+            properties["azfunction_appId"] = process.env.WEBSITE_HOSTNAME;
+        } else if (config) { // VM
+            waiting = true;
+            this._getAzureComputeMetadata(config, () => {
+                if (this._isVM && Object.keys(this._vmData).length > 0) {
+                    properties["azInst_vmId"] = this._vmData["vmId"] || "";
+                    properties["azInst_subscriptionId"] = this._vmData["subscriptionId"] || "";
+                    properties["azInst_osType"] = this._vmData["osType"] || "";
+                }
+                this._client.trackMetric({name: Constants.HeartBeatMetricName, value: 0, properties: properties});
+                callback();
+            });
+        }
+        if (!waiting) {
+            this._client.trackMetric({name: Constants.HeartBeatMetricName, value: 0, properties: properties});
+            callback();
+        }
+    }
+
+    public dispose() {
+        HeartBeat.INSTANCE = null;
+        this.enable(false);
+        this._isInitialized = false;
+    }
+
+    private _getAzureComputeMetadata(config: Config, callback: () => void) {
+        const metadataRequestUrl = `${HeartBeat.AIMS_URI}?${HeartBeat.AIMS_API_VERSION}&${HeartBeat.AIMS_FORMAT}`;
+        const requestOptions = {
+            method: 'GET',
+            headers: {
+                "Metadata": "true"
+            }
+        };
+
+        const req = Util.makeRequest(config, metadataRequestUrl, requestOptions, (res) => {
+            if (res.statusCode === 200) {
+                // Success; VM
+                this._isVM = true;
+                let data = "";
+                res.on('data', (data: any) => {
+                    data += data;
+                });
+                res.on('end', () => {
+                    this._vmData = JSON.parse(data);
+                    callback();
+                });
+            } else if (res.statusCode >= 400 && res.statusCode < 500) {
+                // Not found; Not in VM; Do not try again.
+                this._isVM = false;
+                callback();
+            } else {
+                // else Retry on next heartbeat metrics call
+                callback();
```

Comment:
I'm not sure how `callback` works. Does this give control back to the outer function? Wouldn't that mean that the below logic will never be run?